Mathlib/Combinatorics/Enumerative/DyckWord.lean

```diff
@@ -0,0 +1,431 @@
+      omega
+
+/-- The right part of the Dyck word decomposition. -/
+def rightPart (h : p.l ≠ []) : DyckWord :=
+  p.drop (p.firstReturn + 1) (p.count_eq_count_of_take_firstReturn_add_one h)
+
+/-- The left part of the Dyck word decomposition. -/
+def leftPart (h : p.l ≠ []) : DyckWord :=
+  (p.take (p.firstReturn + 1) (p.count_eq_count_of_take_firstReturn_add_one h)).denest
+    (by simp [take, h]) (fun i lb ub ↦ by
+      simp only [take, length_take, lt_min_iff] at ub ⊢
+      replace ub := ub.1
+      rw [take_take, min_eq_left ub.le]
+      rw [show i = i - 1 + 1 by omega] at ub ⊢
+      rw [Nat.add_lt_add_iff_right] at ub
+      exact p.count_lt_count_of_lt_firstReturn _ ub)
+
+@[simp]
+theorem leftPart_nest_add_rightPart (h : p.l ≠ []) : (p.leftPart h).nest + p.rightPart h = p := by
+  rw [DyckWord.ext_iff, leftPart, denest_nest]
+  apply take_append_drop
+
+lemma leftPart_length_lt (h : p.l ≠ []) : (p.leftPart h).l.length < p.l.length := by
+  nth_rw 2 [← p.leftPart_nest_add_rightPart h]
+  change _ < (U :: (p.leftPart h).l ++ [D] ++ (p.rightPart h).l).length
+  simp_rw [length_append, length_singleton, length_cons]; omega
+
+theorem leftPart_semilength_lt (h : p.l ≠ []) : (p.leftPart h).semilength < p.semilength := by
+  rw [← Nat.mul_lt_mul_left zero_lt_two]
+  simp_rw [two_mul_semilength_eq_length]
+  exact p.leftPart_length_lt h
+
+lemma rightPart_length_lt (h : p.l ≠ []) : (p.rightPart h).l.length < p.l.length := by
+  nth_rw 2 [← p.leftPart_nest_add_rightPart h]
+  change _ < (U :: (p.leftPart h).l ++ [D] ++ (p.rightPart h).l).length
+  simp_rw [length_append, length_singleton, length_cons]; omega
+
+theorem rightPart_semilength_lt (h : p.l ≠ []) : (p.rightPart h).semilength < p.semilength := by
+  rw [← Nat.mul_lt_mul_left zero_lt_two]
+  simp_rw [two_mul_semilength_eq_length]
+  exact p.rightPart_length_lt h
+
+lemma nest_add_ne_empty : (p.nest + q).l ≠ [] := by change U :: _ ++ _ ≠ []; simp
+
+lemma nest_add_firstReturn : (p.nest + q).firstReturn = p.l.length + 1 := by
+  rw [firstReturn_add, firstReturn_nest, ite_eq_right_iff, eq_zero_iff]; intro; contradiction
+
+theorem nest_add_leftPart_eq : (p.nest + q).leftPart (nest_add_ne_empty _ _) = p := by
+  have : p.l.length + 1 + 1 = p.nest.l.length := by simp [nest]
+  simp_rw [leftPart, nest_add_firstReturn, take, DyckWord.ext_iff,
+    show (p.nest + q).l = p.nest.l ++ q.l by rfl, take_append_eq_append_take,
+    this, take_length, tsub_self, take_zero, append_nil, denest, nest, dropLast_concat]
+  rfl
+
+theorem nest_add_rightPart_eq : (p.nest + q).rightPart (nest_add_ne_empty _ _) = q := by
+  have : p.l.length + 1 + 1 = p.nest.l.length := by simp [nest]
+  simp_rw [rightPart, nest_add_firstReturn, drop, DyckWord.ext_iff,
+    show (p.nest + q).l = p.nest.l ++ q.l by rfl, drop_append_eq_append_drop,
+    this, drop_length, nil_append, tsub_self, drop_zero]
+
+end FirstReturn
+
+section Tree
+
+open Tree
+
+/-- Convert a Dyck word to a binary rooted tree. -/
```

Comment:
Can you include somewhere in this file (module doc or doc string) an informal description of the bijection?

Reply:
Added in `DyckWord.treeEquivToFun` and `DyckWord.treeEquivInvFun`.